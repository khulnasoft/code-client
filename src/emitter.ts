import { EventEmitter } from 'events';

import { AnalysisResponseProgress } from './http';
import { SupportedFiles } from './interfaces/files.interface';

// eslint-disable-next-line no-shadow
enum CUSTOM_EVENTS {
  supportedFilesLoaded = 'supportedFilesLoaded',
  scanFilesProgress = 'scanFilesProgress',
  createBundleProgress = 'createBundleProgress',
  uploadBundleProgress = 'uploadBundleProgress',
  analyseProgress = 'analyseProgress',
  apiRequestLog = 'apiRequestLog',
  error = 'error',
}

export class EmitterDC extends EventEmitter {
  events = CUSTOM_EVENTS;

  supportedFilesLoaded(data: SupportedFiles | null): void {
    this.emit(CUSTOM_EVENTS.supportedFilesLoaded, data);
  }

  scanFilesProgress(processed: number): void {
    this.emit(CUSTOM_EVENTS.scanFilesProgress, processed);
  }

  createBundleProgress(processed: number, total: number): void {
    this.emit(CUSTOM_EVENTS.createBundleProgress, processed, total);
  }

  uploadBundleProgress(processed: number, total: number): void {
    this.emit(CUSTOM_EVENTS.uploadBundleProgress, processed, total);
  }

  analyseProgress(data: AnalysisResponseProgress): void {
    this.emit(CUSTOM_EVENTS.analyseProgress, data);
  }

  sendError(error: Error): void {
    this.emit(CUSTOM_EVENTS.error, error);
  }

  apiRequestLog(message: string): void {
    this.emit(CUSTOM_EVENTS.apiRequestLog, message);
  }
}

export const emitter = new EmitterDC();
