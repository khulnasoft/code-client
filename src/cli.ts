/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { Command } from 'commander';

import {
  getSupportedFiles,
  CreateBundleFromFoldersOptions,
  createBundleWithCustomFiles,
  createBundleFromFolders,
  FileBundle,
} from './bundles';
import { analyzeBundle } from './analysis';
import { ConnectionOptions, checkBundle, AnalysisResponseProgress } from './http';
import { emitter } from './emitter';

function parseConnectionOptions(options: {
  url: string;
  token: string;
  source: string;
  org?: string;
  headers?: string[];
}): ConnectionOptions {
  let headers: { [key: string]: string } = {};
  if (options.headers?.length) {
    headers = options.headers.reduce((m, h) => {
      const [key, value] = h.split(':');
      m[key.trim()] = value.trim();
      return m;
    }, {});
  }
  return {
    baseURL: options.url,
    sessionToken: options.token,
    source: options.source,
    ...(options.org ? { org: options.org } : {}),
    extraHeaders: headers,
    // requestId?: string; - not supported
  };
}

async function filtersAction(options: { url: string; token: string; source: string; org?: string }) {
  const response = await getSupportedFiles(options.url, options.source);
  console.log(JSON.stringify(response, null, 2));
}

async function createBundleAction(options: {
  url: string;
  token: string;
  source: string;
  org?: string;
  headers?: string[];
  debug?: boolean;
  ignore?: string[];
  patterns?: string[];
  path: string[];
}) {
  const opts: CreateBundleFromFoldersOptions = {
    ...parseConnectionOptions(options),
    paths: options.path,
    ...(options.ignore?.length ? { defaultFileIgnores: options.ignore } : {}),
  };
  let bundle: FileBundle | null = null;
  if (options.patterns?.length) {
    bundle = await createBundleWithCustomFiles(opts, { configFiles: [], extensions: options.patterns ?? [] });
  } else {
    bundle = await createBundleFromFolders(opts);
  }
  if (bundle === null) {
    process.exitCode = 1;
  }

  if (options.debug) {
    console.log(JSON.stringify(bundle, null, 2));
  } else {
    console.log(bundle?.bundleHash);
  }
}

async function readBundleAction(
  bundleHash: string,
  options: { url: string; token: string; source: string; org?: string },
) {
  const opts = {
    ...parseConnectionOptions(options),
    bundleHash,
  };

  const bundle = await checkBundle(opts);
  if (bundle.type === 'error') {
    process.exitCode = 1;
  }
  console.log(JSON.stringify(bundle, null, 2));
}

async function analysisBundleAction(
  bundleHash: string,
  options: { url: string; token: string; source: string; org?: string },
) {
  const opts = {
    ...parseConnectionOptions(options),
    bundleHash,
  };

  const bundle = await analyzeBundle(opts);
  console.log(JSON.stringify(bundle, null, 2));
}

function enhanceCommonOptions(command: Command): Command {
  return command
    .requiredOption('--url <url>', 'service URL')
    .requiredOption('--token <hash>', 'user token')
    .option('--org <string>', 'organization')
    .option('--source <string>', 'source identifier', 'code-client')
    .option(
      '-H, --headers [string...]',
      'custom headers e.g. "X-Custom-Header: some value". Can have multiple values diveded by space',
    )
    .option('--debug', 'enable debug mode');
}

const program = new Command();

program.name('CLI').version('0.1.0').description('Code Client API');

const filtersCommand = new Command('filters').description('get supported file filters').action(filtersAction);

program.addCommand(enhanceCommonOptions(filtersCommand));

const createCommand = new Command('bundle:create')
  .description('create a new bundle and return its ID with meta info')
  .option('--patterns [string...]', 'supported file patterns')
  .option('--ignore [path...]', 'ignored path glob')
  .requiredOption('--path [path...]', 'source code dir')
  .action(createBundleAction);

program.addCommand(enhanceCommonOptions(createCommand));

const checkCommand = new Command('bundle:check')
  .description('read existing bundle and return its ID with meta info')
  .argument('<bundleId>', 'bundle identifier')
  .action(readBundleAction);

program.addCommand(enhanceCommonOptions(checkCommand));

const analysisCommand = new Command('analysis:bundle')
  .description('run analysis for existing bundle')
  .argument('<bundleId>', 'bundle identifier')
  .action(analysisBundleAction);

program.addCommand(enhanceCommonOptions(analysisCommand));

/** Bundle upload process is started with provided data */
emitter.on('uploadBundleProgress', (processed: number, total: number) => {
  console.log(`Upload bundle progress: ${processed}/${total}`);
});

emitter.on('analyseProgress', (data: AnalysisResponseProgress) => {
  console.log(`${data.status} : ${data.progress * 100}%`);
});

// emitter.on('apiRequestLog', (message: string) => {
//   console.log(message);
// });

async function main(): Promise<void> {
  await program.parseAsync(process.argv);
}

main().catch(err => {
  console.error(err);
});
