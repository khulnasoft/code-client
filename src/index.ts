import { analyzeFolders, extendAnalysis, analyzeScmProject } from './analysis';
import { getSupportedFiles, createBundleFromFolders, createBundleWithCustomFiles } from './bundles';
import { emitter } from './emitter';
import { startSession, checkSession, getAnalysis, getIpFamily, IpFamily } from './http';
import { MAX_FILE_SIZE } from './constants';
import * as constants from './constants';
import { getGlobPatterns } from './files';

import { SupportedFiles, FileAnalysis } from './interfaces/files.interface';
import { AnalysisSeverity, AnalysisContext } from './interfaces/analysis-options.interface';
import {
  AnalysisResult,
  AnalysisResultSarif,
  AnalysisResultLegacy,
  FilePath,
  FileSuggestion,
  Suggestion,
  Marker,
  ReportResult,
  ScmAnalysis,
} from './interfaces/analysis-result.interface';

export {
  getGlobPatterns,
  analyzeFolders,
  analyzeScmProject,
  getSupportedFiles,
  createBundleFromFolders,
  createBundleWithCustomFiles,
  extendAnalysis,
  emitter,
  MAX_FILE_SIZE,
  constants,
  AnalysisSeverity,
  AnalysisResult,
  AnalysisResultSarif,
  AnalysisResultLegacy,
  SupportedFiles,
  FileAnalysis,
  FilePath,
  FileSuggestion,
  Suggestion,
  Marker,
  getAnalysis,
  startSession,
  checkSession,
  getIpFamily,
  IpFamily,
  AnalysisContext,
  ReportResult,
  ScmAnalysis,
};
