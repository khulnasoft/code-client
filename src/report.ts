import * as uuid from 'uuid';
import pick from 'lodash.pick';
import { POLLING_INTERVAL } from './constants';
import { emitter } from './emitter';
import {
  AnalysisFailedResponse,
  GetAnalysisErrorCodes,
  UploadReportResponseDto,
  initReport,
  initScmReport,
  AnalysisStatus,
  Result,
  UploadReportOptions,
  getReport,
  getScmReport,
  ScmUploadReportOptions,
  GetReportOptions,
} from './http';
import { ReportResult } from './interfaces/analysis-result.interface';

const sleep = (duration: number) => new Promise(resolve => setTimeout(resolve, duration));

type InitReportGenericType = (
  options: ScmUploadReportOptions | UploadReportOptions,
) => Promise<Result<string, GetAnalysisErrorCodes>>;

type GetReportGenericType = (
  options: GetReportOptions,
) => Promise<Result<UploadReportResponseDto, GetAnalysisErrorCodes>>;

type ReportOptionsGenericType = UploadReportOptions | ScmUploadReportOptions;

async function initAndPollReportGeneric(
  initReportFunc: InitReportGenericType,
  getReportFunc: GetReportGenericType,
  options: ReportOptionsGenericType,
) {
  emitter.analyseProgress({
    status: AnalysisStatus.waiting,
    progress: 0,
  });

  // First init the report
  const initResponse = await initReportFunc(options);

  if (initResponse.type === 'error') {
    return initResponse;
  }
  const pollId = initResponse.value;

  let apiResponse: Result<UploadReportResponseDto, GetAnalysisErrorCodes>;
  let response: UploadReportResponseDto;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    // eslint-disable-next-line no-await-in-loop
    apiResponse = await getReportFunc({
      ...pick(options, ['baseURL', 'sessionToken', 'source', 'requestId', 'org']),
      pollId,
    });

    if (apiResponse.type === 'error') {
      return apiResponse;
    }

    response = apiResponse.value;

    if (
      response.status === AnalysisStatus.waiting ||
      response.status === AnalysisStatus.fetching ||
      response.status === AnalysisStatus.analyzing ||
      response.status === AnalysisStatus.done
    ) {
      // Report progress of fetching
      emitter.analyseProgress(response);
    } else if (response.status === AnalysisStatus.complete) {
      // Return data of analysis
      return apiResponse as Result<ReportResult, GetAnalysisErrorCodes>;
      // deepcode ignore DuplicateIfBody: false positive it seems that interface is not taken into account
    } else if (response.status === AnalysisStatus.failed) {
      // Report failure of analysing
      return apiResponse as Result<AnalysisFailedResponse, GetAnalysisErrorCodes>;
    }

    // eslint-disable-next-line no-await-in-loop
    await sleep(POLLING_INTERVAL);
  }
}

export async function reportBundle(options: UploadReportOptions): Promise<ReportResult> {
  const projectName = options.report?.projectName?.trim();
  if (!projectName || projectName.length === 0) {
    throw new Error('"project-name" must be provided for "report"');
  }
  if (/[^A-Za-z0-9-_/]/g.test(projectName)) {
    throw new Error(`"project-name" must not contain spaces or special characters except [/-_]`);
  }

  // Trigger bundle analysis and emit intermediate progress.
  const response = await initAndPollReportGeneric(initReport, getReport, options);

  if (response.type === 'error') {
    throw response.error;
  } else if (response.value.status === AnalysisStatus.failed) {
    throw new Error('Analysis has failed');
  }

  return response.value;
}

export async function reportScm(options: ScmUploadReportOptions): Promise<ReportResult> {
  const projectId = options.projectId?.trim();
  if (!projectId || projectId.length === 0) {
    throw new Error('"project-id" must be provided for "report"');
  }
  if (!uuid.validate(projectId)) {
    throw new Error('"project-id" must be a valid UUID');
  }

  const commitId = options.commitId?.trim();
  if (!commitId || commitId.length === 0) {
    throw new Error('"commit-id" must be provided for "report"');
  }

  // Trigger SCM project analysis and emit intermediate progress.
  const response = await initAndPollReportGeneric(initScmReport, getScmReport, options);

  if (response.type === 'error') {
    throw response.error;
  } else if (response.value.status === AnalysisStatus.failed) {
    throw new Error('Analysis has failed');
  }

  return response.value;
}
