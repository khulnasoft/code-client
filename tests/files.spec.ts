import * as fs from 'fs';
import * as nodePath from 'path';
import 'jest-extended';

import {
  collectBundleFiles,
  prepareExtendingBundle,
  composeFilePayloads,
  getFileInfo,
  getBundleFilePath,
  resolveBundleFilePath,
  collectFilePolicies,
} from '../src/files';
import { getGlobPatterns } from '../src';
import { FileInfo } from '../src/interfaces/files.interface';
import {
  sampleProjectPath,
  supportedFiles,
  bundleFiles,
  bundleFilesFull,
  bundleFileIgnores,
  fileIgnoresFixtures,
  bundleFilePolicies,
} from './constants/sample';

describe('files', () => {
  it('collect bundle files', async () => {
    // TODO: We should introduce some performance test using a big enough repo to avoid flaky results
    const collector = collectBundleFiles({
      baseDir: sampleProjectPath,
      paths: [sampleProjectPath],
      supportedFiles,
      filePolicies: bundleFilePolicies,
    });
    const files: FileInfo[] = [];
    const skippedOversizedFiles: string[] = [];
    for await (const f of collector) {
      typeof f == 'string' ? skippedOversizedFiles.push(f) : files.push(f);
    }
    // all files in the repo are expected other than the file that exceeds MAX_FILE_SIZE 'big-file.js'
    const expectedFiles = (await bundleFiles).filter(obj => !obj.bundlePath.includes('big-file.js'));
    expect(files.map(f => f.filePath)).toIncludeSameMembers(expectedFiles.map(f => f.filePath)); // Assert same files in bundle
    expect(files).toIncludeSameMembers(expectedFiles); // Assert same properties for files in bundle

    // big-file.js should be added to skippedOversizedFiles
    expect(skippedOversizedFiles.length).toEqual(1);
    expect(skippedOversizedFiles[0]).toEqual('big-file.js');

    const filesWithoutBasePath = files.map(f => ({
      ...f,
      filePath: f.filePath.replace(sampleProjectPath, '<basePath>'),
    }));
    expect(filesWithoutBasePath).toMatchSnapshot();
  });

  it('collects only non-excluded files', async () => {
    const testPath = `${fileIgnoresFixtures}/negative-overrides`;
    const filePolicies = await collectFilePolicies([testPath]);
    const collector = collectBundleFiles({
      baseDir: testPath,
      paths: [testPath],
      supportedFiles,
      filePolicies,
    });
    const files: FileInfo[] = [];
    for await (const f of collector) {
      typeof f !== 'string' && files.push(f);
    }

    expect(files).toHaveLength(1);
    expect(files[0].bundlePath).toBe('.khulnasoft');
  });

  it('extend bundle files', async () => {
    const testNewFiles = [`app.js`, `not/ignored/this_should_not_be_ignored.java`];
    const testRemovedFiles = [`removed_from_the_parent_bundle.java`];
    const newBundle = [...testNewFiles, ...testRemovedFiles];
    const { files, removedFiles } = await prepareExtendingBundle(
      sampleProjectPath,
      supportedFiles,
      bundleFileIgnores,
      newBundle,
    );
    expect(files).toEqual((await bundleFiles).filter(obj => testNewFiles.includes(obj.bundlePath)));
    expect(removedFiles).toEqual(['removed_from_the_parent_bundle.java']);
  });

  it('collect bundle files with multiple folders', async () => {
    // Limit size and we get fewer files
    const folders = [nodePath.join(sampleProjectPath, 'models'), nodePath.join(sampleProjectPath, 'controllers')];
    const collector = collectBundleFiles({
      baseDir: sampleProjectPath,
      paths: folders,
      supportedFiles,
      filePolicies: {
        excludes: [],
        ignores: [],
      },
    });
    const smallFiles: FileInfo[] = [];
    const skippedOversizedFiles: string[] = [];
    for await (const f of collector) {
      typeof f == 'string' ? skippedOversizedFiles.push(f) : smallFiles.push(f);
    }
    expect(smallFiles.length).toEqual(2);
    expect(skippedOversizedFiles.length).toEqual(0);
    expect(smallFiles.map(f => [f.filePath, f.bundlePath])).toEqual([
      [`${sampleProjectPath}/models/sharks.js`, 'models/sharks.js'],
      [`${sampleProjectPath}/controllers/sharks.js`, 'controllers/sharks.js'],
    ]);
  });

  it('compose file payloads', async () => {
    // Prepare all missing files first
    const payloads = [...composeFilePayloads(await bundleFilesFull, 1024)];
    expect(payloads.length).toEqual(4); // 4 chunks
    expect(payloads[0].length).toEqual(4); // 4 files in first chunk

    const testPayload = payloads[3][0];
    expect(testPayload.filePath).toEqual(`${sampleProjectPath}/routes/sharks.js`);
    expect(testPayload.bundlePath).toEqual(`routes/sharks.js`);
    expect(testPayload.size).toEqual(363);
    expect(testPayload.hash).toEqual('f870a225bfad387cd3c46ccfb0be52415aa2e07767b53edae54c41fa8e12e82e');
    expect(testPayload.content).toEqual(fs.readFileSync(testPayload.filePath).toString('utf8'));
  });

  it('support of utf-8 encoding', async () => {
    // fs.readFileSync(payloads[0][0].path).toString('utf8')
    const filePath = `${sampleProjectPath}/app.js`;
    const fileMeta = await getFileInfo(filePath, sampleProjectPath);
    expect(fileMeta?.hash).toEqual('40f937553fda7b9986c3a87d39802b96e77fb2ba306dd602f9b2d28949316c98');
  });

  it('support of iso8859 encoding', async () => {
    const filePath = `${sampleProjectPath}/main.js`;
    const fileMeta = await getFileInfo(filePath, sampleProjectPath);
    expect(fileMeta?.hash).toEqual('3e2979852cc2e97f48f7e7973a8b0837eb73ed0485c868176bc3aa58c499f534');
  });

  it('gets correct bundle file path if no baseDir specified', () => {
    const baseDir = '';
    const darwinPath = '/Users/user/Git/goof/routes/index.js';
    expect(getBundleFilePath(darwinPath, baseDir)).toEqual(darwinPath);

    const linuxPath = '/home/user/Git/goof/routes/index.js';
    expect(getBundleFilePath(linuxPath, baseDir)).toEqual(linuxPath);

    const windowsPath = 'C:\\Users\\user\\Git\\goof\\index.js';
    expect(getBundleFilePath(windowsPath, baseDir)).toEqual(encodeURI(windowsPath));
  });

  it('resolves correct bundle file path if no baseDir specified and contains whitespace', () => {
    const baseDir = '';
    const darwinPath = '/Users/user/Git/goof%20test/routes/index.js';
    expect(resolveBundleFilePath(baseDir, darwinPath)).toEqual(decodeURI(darwinPath));

    const linuxPath = '/home/user/Git/goof%20test/routes/index.js';
    expect(resolveBundleFilePath(baseDir, linuxPath)).toEqual(decodeURI(linuxPath));

    const windowsPath = 'C:\\Users\\user\\Git\\goof%20test\\index.js';
    expect(resolveBundleFilePath(baseDir, windowsPath)).toEqual(decodeURI(windowsPath));
  });

  describe('getGlobPatterns', () => {
    it('generates correct glob patterns for supported files', () => {
      const globPatterns = getGlobPatterns(supportedFiles);
      expect(globPatterns).toEqual([
        '*.[jJ][sS]',
        '*.[jJ][sS][xX]',
        '*.[cC][pP][pP]',
        '*.[jJ][aA][vV][aA]',
        '.eslintrc.json',
        '.khulnasoft',
      ]);
    });

    it("doesn't generate any pattern for an empty file extension", () => {
      const supportedFiles = {
        extensions: ['', '.cs'],
        configFiles: [],
      };
      const globPatterns = getGlobPatterns(supportedFiles);
      expect(globPatterns).toEqual(['', '*.[cC][sS]']);
    });

    it("doesn't generate any pattern for invalid file extension", () => {
      const supportedFiles = {
        extensions: ['js', '.cs'],
        configFiles: [],
      };
      const globPatterns = getGlobPatterns(supportedFiles);
      expect(globPatterns).toEqual(['', '*.[cC][sS]']);
    });

    it("doesn't generate case variant pattern for chars without case variant", () => {
      const supportedFiles = {
        extensions: ['.ps1', '.ps'],
        configFiles: [],
      };
      const globPatterns = getGlobPatterns(supportedFiles);
      expect(globPatterns).toEqual(['*.[pP][sS]1', '*.[pP][sS]']);
    });
  });
});
