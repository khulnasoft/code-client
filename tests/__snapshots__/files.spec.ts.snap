// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files collect bundle files 1`] = `
Array [
  Object {
    "bundlePath": ".eslintrc.json",
    "content": undefined,
    "filePath": "<basePath>/.eslintrc.json",
    "hash": "d3332248171b63cf6f785a3750a27bf1cd03bc0462458943d937c34494f08106",
    "size": 186,
  },
  Object {
    "bundlePath": ".khulnasoft",
    "content": undefined,
    "filePath": "<basePath>/.khulnasoft",
    "hash": "92f1200f08dba7ec777a8499f5a6d8ac63d5c3d6ddca58b2069a840fd0132790",
    "size": 92,
  },
  Object {
    "bundlePath": "AnnotatorTest.Cpp",
    "content": undefined,
    "filePath": "<basePath>/AnnotatorTest.Cpp",
    "hash": "61b028b49c2a4513b1c7c161b5f491264fe71c9c29bc0ae8e6d760c156b45edc",
    "size": 239,
  },
  Object {
    "bundlePath": "GitHubAccessTokenScrambler12.java",
    "content": undefined,
    "filePath": "<basePath>/GitHubAccessTokenScrambler12.java",
    "hash": "63184d2adbf4a3fe17430f1029eb9d7c0aaa8cf8a00f20199b921175c0f30d2b",
    "size": 949,
  },
  Object {
    "bundlePath": "app.js",
    "content": undefined,
    "filePath": "<basePath>/app.js",
    "hash": "40f937553fda7b9986c3a87d39802b96e77fb2ba306dd602f9b2d28949316c98",
    "size": 510,
  },
  Object {
    "bundlePath": "db.js",
    "content": undefined,
    "filePath": "<basePath>/db.js",
    "hash": "6f8d7925b5c86bd6d31b0b23bdce1dcfc94e28a1d5ebdc0ba91fac7dc7e95657",
    "size": 533,
  },
  Object {
    "bundlePath": "main.js",
    "content": undefined,
    "filePath": "<basePath>/main.js",
    "hash": "3e2979852cc2e97f48f7e7973a8b0837eb73ed0485c868176bc3aa58c499f534",
    "size": 23098,
  },
  Object {
    "bundlePath": "exclude/.khulnasoft",
    "content": undefined,
    "filePath": "<basePath>/exclude/.khulnasoft",
    "hash": "b33a7db205b03ccb488773ea895d70b6a777366ae850d2f4e3bcec7293831d71",
    "size": 45,
  },
  Object {
    "bundlePath": "routes/index.js",
    "content": undefined,
    "filePath": "<basePath>/routes/index.js",
    "hash": "86ad66d869e6a1bfb5edbc6a82203eece06db4690e4480b3b69274a5c91052bd",
    "size": 300,
  },
  Object {
    "bundlePath": "routes/sharks.js",
    "content": undefined,
    "filePath": "<basePath>/routes/sharks.js",
    "hash": "f870a225bfad387cd3c46ccfb0be52415aa2e07767b53edae54c41fa8e12e82e",
    "size": 363,
  },
  Object {
    "bundlePath": "not/ignored/this_should_be_ignored.jsx",
    "content": undefined,
    "filePath": "<basePath>/not/ignored/this_should_be_ignored.jsx",
    "hash": "3dce2c0456d5f9b6ff64c589047f359b0a7ec390ec1edd20b0070d3399e4e96a",
    "size": 252,
  },
  Object {
    "bundlePath": "not/ignored/this_should_not_be_ignored.java",
    "content": undefined,
    "filePath": "<basePath>/not/ignored/this_should_not_be_ignored.java",
    "hash": "63184d2adbf4a3fe17430f1029eb9d7c0aaa8cf8a00f20199b921175c0f30d2b",
    "size": 949,
  },
]
`;
